pkg/reconciler/namespace/namespace.go

```diff
@@ -100,55 +103,156 @@ func (c *Controller) reconcileGVR(ctx context.Context, gvr schema.GroupVersionRe
+		newClusterName = cluster.Name
+	}
 
-		// Patch the namespace.
-		klog.Infof("Patching to update cluster assignment for namespace %s: %q -> %q", ns.Name, old, new)
-		if _, err := c.namespaceClient.Patch(ctx, ns.Name, types.MergePatchType, clusterLabelPatchBytes(new), metav1.PatchOptions{}); err != nil {
+	if oldClusterName != newClusterName {
+		klog.Infof("Patching to update cluster assignment for namespace %s: %q -> %q", ns.Name, oldClusterName, newClusterName)
+		if _, err := c.namespaceClient.Patch(ctx, ns.Name, types.MergePatchType, clusterLabelPatchBytes(newClusterName), metav1.PatchOptions{}); err != nil {
 			return err
 		}
+
 	}
 
-	// Check if the cluster reports as Ready.
-	cl, err := c.clusterLister.Get(clusters.ToClusterAwareKey(lclusterName, clusterName))
+	return nil
+}
+
+// isValidCluster checks whether the given cluster name exists and is valid for the
+// purposes of scheduling a namespace to it.
+func (c *Controller) isValidCluster(ctx context.Context, lclusterName, clusterName string) (bool, error) {
+	cluster, err := c.clusterLister.Get(clusters.ToClusterAwareKey(lclusterName, clusterName))
+	if apierrors.IsNotFound(err) {
+		klog.Infof("Cluster %q does not exist", clusterName)
+		return false, nil
+	}
 	if err != nil {
+		return false, err
+	}
+	isReady := cluster.Status.Conditions.HasReady()
+	if !isReady {
+		klog.Infof("Cluster %q is not reporting ready", cluster.Name)
+	}
+	return isReady, nil
+}
+
+// ensureScheduled attempts to ensure the namespace is assigned to a viable cluster. This
+// will succeed without error if a cluster is assigned or if there are no viable clusters
+// to assign to. The condition of not being scheduled to a cluster will be reflected in
+// the namespace's status rather than by returning an error.
+func (c *Controller) ensureScheduled(ctx context.Context, lclusterName string, ns *corev1.Namespace) error {
+	clusterName := ns.Labels[clusterLabel]
+
+	noClusterAssigned := clusterName == ""
+	if noClusterAssigned {
+		if err := c.assignCluster(ctx, ns); err != nil {
+			return err
+		}
+	} else {
+		isValidCluster, err := c.isValidCluster(ctx, lclusterName, clusterName)
+		if err != nil {
+			return err
+		}
+		if !isValidCluster {
+			// Currently assigned cluster no longer exists or is not ready.
+			if err := c.assignCluster(ctx, ns); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// ensureScheduledStatus ensures the status of the given namespace reflects whether the
+// namespace has been assigned to a cluster.
+func (c *Controller) ensureScheduledStatus(ctx context.Context, ns *corev1.Namespace, oldClusterName string) error {
+	// Set status on cluster change to bump the transition time
+	newClusterName := ns.Labels[clusterLabel]
+	clusterChanged := oldClusterName != newClusterName
+
+	// Set status if the scheduled state of the namespace has changed
+	scheduled := (newClusterName != "")
+	statusChangeRequired := scheduled != HasScheduledStatus(ns)
```

Comment:
So one of the benefits of the library I *think* is that attempting to set a condition to a value that's already set will be a no-op, so we should not have to do this sort of check?

Reply:
afaict there is nothing preventing us making a call to the server without these checks. What am I missing?

Reply:
@ncdc might know -- but since you're building your own conditions adapter anyway it may not matter. I was under the impression that one of the big wins for the library was that it would handle deduping/re-setting.

Reply:
Maybe you're thinking about the library setting the condition on a resource locally? Persisting that change still requires a server call (e.g. via `Patch`) and afaict there's no way to avoid that call without the kind of check you see here.

Reply:
I assumed that the double-JSON-serialization and JSON Patch logic would be able to detect the no-op and short-circuit the call, so the code here (and elsewhere) can focus on declaration of intent - not imperative update-if-not-set logic

Reply:
@sttts Before I dig into code, does `Patch` detect no-op and thus avoid making unnecessary api calls?